{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimize the primer regions used"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code to find the subsequences of the given primers for which the greatest number of sequences can be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modified versions of the functions used to extract the region between two primers, \n",
    "# by doing a pairwise alignment to a referense sequence for which the primer positions\n",
    "# are known.\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import PairwiseAligner\n",
    "import regex as re\n",
    "from itertools import product\n",
    "\n",
    "def align_to_reference(query_seq, reference_seq):\n",
    "    \"\"\"\n",
    "    Align the query sequence to the reference sequence and return alignment details.\n",
    "    \"\"\"\n",
    "    aligner = PairwiseAligner()\n",
    "    aligner.mode = 'global'\n",
    "    aligner.match_score = 1\n",
    "    aligner.mismatch_score = -1\n",
    "    aligner.open_gap_score = -2\n",
    "    aligner.extend_gap_score = -1\n",
    "\n",
    "    alignment = aligner.align(reference_seq, query_seq)[0]  # Get the best alignment\n",
    "    aligned_ref = alignment.aligned[0]\n",
    "    aligned_query = alignment.aligned[1]\n",
    "    \n",
    "    return aligned_ref, aligned_query\n",
    "\n",
    "def extract_region_of_interest(aligned_ref, aligned_query, ref_start, ref_end):\n",
    "    \"\"\"\n",
    "    Extract the region of interest from the aligned query sequence based on reference coordinates.\n",
    "    \"\"\"\n",
    "    query_start = None\n",
    "    query_end = None\n",
    "\n",
    "    # Map reference positions to query sequence positions\n",
    "    for ref_range, query_range in zip(aligned_ref, aligned_query):\n",
    "        # Check if the primer region falls within this reference range\n",
    "        if ref_start >= ref_range[0] and ref_start < ref_range[1]:\n",
    "            query_start = query_range[0] + (ref_start - ref_range[0])\n",
    "        if ref_end > ref_range[0] and ref_end <= ref_range[1]:\n",
    "            query_end = query_range[0] + (ref_end - ref_range[0])\n",
    "    \n",
    "    return query_start, query_end\n",
    "\n",
    "def process_sequences(fasta_file, reference_seq, primer_fwd, primer_rev):\n",
    "    \"\"\"\n",
    "    Extract sequences between primers based on alignment to a reference sequence.\n",
    "    \"\"\"\n",
    "    # Find positions of primers in the reference sequence\n",
    "    ref_start = reference_seq.index(primer_fwd) + len(primer_fwd)\n",
    "    ref_end = reference_seq.index(primer_rev)\n",
    "\n",
    "    not_found = []\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        seq = str(record.seq).upper()\n",
    "\n",
    "        # Align sequence to reference\n",
    "        aligned_ref, aligned_query = align_to_reference(seq, reference_seq)\n",
    "\n",
    "        start, end = extract_region_of_interest(seq, aligned_ref, aligned_query, ref_start, ref_end)\n",
    "\n",
    "        if start is None or end is None:\n",
    "            not_found.append(record.id)\n",
    "    \n",
    "    return not_found\n",
    "\n",
    "def find_primers(sequence, primer, length):\n",
    "    '''\n",
    "    Find all possible subsequences of a given length from a primer, \n",
    "    that only exist once in the reference organism\n",
    "    Input: Reference sequence, primer, length of subsequence primers\n",
    "    Output: List of all possible subsequence primers that are found just once in the reference\n",
    "    '''\n",
    "    out_list = []\n",
    "    for i in range(len(primer)-length):\n",
    "        primer_subseq = primer[i:i+length]\n",
    "        matches = re.findall(primer_subseq, sequence)\n",
    "        if len(matches) == 1:\n",
    "            out_list.append(primer_subseq)\n",
    "    \n",
    "    return out_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the reference sequence and primers\n",
    "for record in SeqIO.parse(\"Ecoli.fasta\", \"fasta\"):\n",
    "    ecoli = str(record.seq)\n",
    "\n",
    "# Primers from extract_sequences_16s\n",
    "v1_full = 'GTAACAGGAAGAAGCTTGCTTCTTTGCTGAC'\n",
    "v9_full = 'GTAGGTAGCTTAACCTTCGGGAGGGCGCTTA'\n",
    "\n",
    "v3_full = 'GTACTTTCAGCGGGGAGGAAGGGAGTAAAGTTAATACCTTTGCTCATTGACGTTACCCGCAGAAG'\n",
    "v4_full = 'CGCACGCAGGCGGUUUGUUAAGUCAGAUGUGAAAUCCCCGGGCUCAACCUGGGAACUGCAUCUGAUACUGGCAAGCUUGAGUCUCGUAGAGGGGGGUAGAAUUCCAG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find all possible subsequence primers of lengths 4, 5, and 6\n",
    "# v1\n",
    "v1_four = find_primers(ecoli, v1_full, 4)\n",
    "v1_five = find_primers(ecoli, v1_full, 5)\n",
    "v1_six = find_primers(ecoli, v1_full, 6)\n",
    "v1_all = v1_four + v1_five + v1_six\n",
    "\n",
    "# v9\n",
    "v9_four = find_primers(ecoli, v9_full, 4)\n",
    "v9_five = find_primers(ecoli, v9_full, 5)\n",
    "v9_six = find_primers(ecoli, v9_full, 6)\n",
    "v9_all = v9_four + v9_five + v9_six\n",
    "\n",
    "# v3\n",
    "v3_four = find_primers(ecoli, v3_full[:round(len(v3_full)/2)], 4)\n",
    "v3_five = find_primers(ecoli, v3_full[:round(len(v3_full)/2)], 5)\n",
    "v3_six = find_primers(ecoli, v3_full[:round(len(v3_full)/2)], 6)\n",
    "v3_all = v3_four + v3_five + v3_six\n",
    "\n",
    "# v4\n",
    "v4_four = find_primers(ecoli, v4_full[round(len(v4_full)/2):], 4)\n",
    "v4_five = find_primers(ecoli, v4_full[round(len(v4_full)/2):], 5)\n",
    "v4_six = find_primers(ecoli, v4_full[round(len(v4_full)/2):], 6)\n",
    "v4_all = v4_four + v4_five + v4_six"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through all possible combinations for the v1-v9 region\n",
    "# Save the primer combination and the number of sequences in which the\n",
    "# region of interest cannot be found using that specific primer pair\n",
    "\n",
    "primers_errors_v1v9 = {}\n",
    "\n",
    "for v1, v9 in list(product(v1_all, v9_all)):\n",
    "    errors = process_sequences(\"Vaginal_species_all.fasta\", ecoli, v1, v9)\n",
    "    primers_errors_v1v9[(v1, v9)] = len(errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same but for the v3-v4 region\n",
    "\n",
    "primers_errors_v3v4 = {}\n",
    "\n",
    "for v3, v4 in list(product(v3_all, v4_all)):\n",
    "    errors = process_sequences(\"Vaginal_species_all.fasta\", ecoli, v3, v4)\n",
    "    primers_errors_v3v4[(v3, v4)] = len(errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the results to files\n",
    "import csv\n",
    "\n",
    "with open('optimized_v1v9.csv', \"w\") as out_file:\n",
    "    writer = csv.writer(out_file)\n",
    "    for key, value in primers_errors_v1v9.items():\n",
    "       writer.writerow([key, value])\n",
    "\n",
    "with open('optimized_v3v4.csv', \"w\") as out_file:\n",
    "    writer = csv.writer(out_file)\n",
    "    for key, value in primers_errors_v3v4.items():\n",
    "       writer.writerow([key, value])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
